File: core/src/main/scala/com/myGames/PokemonInfoScala/Player.scala
package com.myGames.PokemonInfoScala

import com.myGames.PokemonInfoScala.Potions.Potions

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

object Player {
  val pokemonBag: ArrayBuffer[AllyPokemon] = ArrayBuffer.empty
  val potionsBag: mutable.Map[Potions, Int] = mutable.Map.empty

  def addPokemon(allyPokemon: AllyPokemon): Unit = {
    pokemonBag += allyPokemon
  }

  def loadPokemon(allyPokemon: List[AllyPokemon]): Unit = {
    pokemonBag.clear()
    pokemonBag ++= allyPokemon
  }

  def addPotion(potion: Potions): Unit = {
    if (potionsBag.contains(potion)) potionsBag += potion -> (potionsBag(potion) + 1)
    else potionsBag += potion -> 1
  }

  def loadPotions(potions: Map[Potions, Int]): Unit = {
    potionsBag.clear()
    potionsBag ++= potions
  }

  def usedPotion(potion: Potions): Unit = {
    if (potionsBag.contains(potion)) {
      val usedValue = potionsBag(potion) - 1
      if (usedValue <= 0) potionsBag -= potion
      else potionsBag += potion -> (potionsBag(potion) - 1)
    }
    else {
      throw new Exception("Potion not found in potion bag.")
    }
  }
}

File: core/src/com/myGames/core/Pokemon/Player.java
package com.myGames.core.Pokemon;

import java.util.HashMap;
import java.util.Map;

public class Player {

    public static final int MAX_POKEMON = 6;
    private static Player INSTANCE = null;
    private AllyPokemon[] pokemonBag;
    private final Map<Potions, Integer> potionsBag;

    private Player() {
        pokemonBag = new AllyPokemon[MAX_POKEMON];
        potionsBag = new HashMap<>();
    }

    public AllyPokemon[] getPokemonBag() {
        return pokemonBag;
    }

    public Map<Potions, Integer> getPotionsBag() {
        return potionsBag;
    }

    public static Player getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new Player();
        }
        return INSTANCE;
    }

    public void addPokemon(AllyPokemon pokemon) {
        // Fill first empty slot with pokemon
        for (int index = 0; index < pokemonBag.length; index++) {
            if (pokemonBag[index] == null) {
                pokemonBag[index] = pokemon;
                return;
            }
        }
    }

    public void loadPokemon(AllyPokemon[] loadedPokemon) {
        pokemonBag = loadedPokemon;
    }

    public void addPotion(Potions potion, int quantity) {
        potionsBag.merge(potion, quantity, Integer::sum);
    }

    public void usePotion(Potions potion) {
        potionsBag.replace(potion, potionsBag.get(potion) - 1);
    }

    public void reset() {
        pokemonBag = new AllyPokemon[MAX_POKEMON];
        potionsBag.clear();
    }
}
